phoenix-core/src/main/java/org/apache/phoenix/coprocessor/CDCGlobalIndexRegionScanner.java

```diff
@@ -223,4 +186,126 @@ protected boolean getNextCoveredIndexRow(List<Cell> result) throws IOException {
+            Map<ImmutableBytesPtr, Cell> changeImageObj,
+            boolean isIndexCellDeleteRow, Long indexCellTS, Cell firstCell) {
+        Map<String, Object> rowValueMap = new HashMap<>();
+
+        Map<String, Object> preImage = new HashMap<>();
+        if (this.cdcChangeScopeSet.size() == 0
+                || (this.cdcChangeScopeSet.contains(PTable.CDCChangeScope.PRE))) {
+            for (Map.Entry<ImmutableBytesPtr, Cell> preImageObjCell : preImageObj.entrySet()) {
+                if (dataColQualNameMap.get(preImageObjCell.getKey()) != null) {
+                    preImage.put(dataColQualNameMap.get(preImageObjCell.getKey()),
+                            dataColQualTypeMap.get(preImageObjCell.getKey()).toObject(
+                                    preImageObjCell.getValue().getValueArray()));
+                }
+            }
+            rowValueMap.put(PRE_IMAGE, preImage);
+        }
+
+        Map<String, Object> changeImage = new HashMap<>();
+        if (this.cdcChangeScopeSet.size() == 0
+                || (this.cdcChangeScopeSet.contains(PTable.CDCChangeScope.CHANGE))) {
+            for (Map.Entry<ImmutableBytesPtr, Cell> changeImageObjCell
+                    : changeImageObj.entrySet()) {
+                if (dataColQualNameMap.get(changeImageObjCell.getKey()) != null) {
+                    changeImage.put(dataColQualNameMap.get(changeImageObjCell.getKey()),
+                            dataColQualTypeMap.get(changeImageObjCell.getKey()).toObject(
+                                    changeImageObjCell.getValue().getValueArray()));
+                }
+            }
+            rowValueMap.put(CHANGE_IMAGE, changeImage);
+        }
+
+        Map<String, Object> postImage = new HashMap<>();
+        if (this.cdcChangeScopeSet.size() == 0
+                || (this.cdcChangeScopeSet.contains(PTable.CDCChangeScope.POST))) {
+            if (!isIndexCellDeleteRow) {
+                for (Map.Entry<ImmutableBytesPtr, Cell> preImageObjCell
+                        : preImageObj.entrySet()) {
+                    if (dataColQualNameMap.get(preImageObjCell.getKey()) != null) {
+                        postImage.put(dataColQualNameMap.get(preImageObjCell.getKey()),
+                                dataColQualTypeMap.get(preImageObjCell.getKey()).toObject(
+                                        preImageObjCell.getValue().getValueArray()));
+                    }
+                }
+                for (Map.Entry<ImmutableBytesPtr, Cell> changeImageObjCell
+                        : changeImageObj.entrySet()) {
+                    if (dataColQualNameMap.get(changeImageObjCell.getKey()) != null) {
+                        postImage.put(dataColQualNameMap.get(changeImageObjCell.getKey()),
+                                dataColQualTypeMap.get(changeImageObjCell.getKey()).toObject(
+                                        changeImageObjCell.getValue().getValueArray()));
+                    }
+                }
+            }
```

Comment:
Instead of using a separate loop for for building postImage, why not do it as part of both the pre and change image? I mean, you would repeat the lines 201 and 215 for postImage as well.